Logging with code cleanup

Greonyral opened the pull request:
* Replace System.err.println by calls to Logger (with internationalized logging messages)
* Decouple classes by interfaces
* Fix typos
* Introduce `@NotNull` and `@Nullable` annotation
* Add `WeakReference` for 'reversed' linking of schedule with train

Greonyral commented on src/main/java/org/gleisbelegung/sts/TrainDetails.java:
```diff
@@ -0,0 +1,184 @@
+    private
+    String target;
+    @Nullable
+    private
```
ja, sieht danach aus => 3962c25

Greonyral commented on src/main/java/org/gleisbelegung/sts/ScheduleFlags.java:
```diff
@@ -1,248 +1,322 @@
+
+    public static final ScheduleFlags EMPTY = new ScheduleFlags();
+    @SuppressWarnings({"unused", "FieldCanBeLocal"})
+    private boolean r;
```
Ja, das wollte ich jetzt aber nicht mehr in den branch reinpacken. Die Abbildung STS-Flag sprechender Name habe ich mit [XmlKeys](https://github.com/manuel3108/Gleisbelegung/blob/9bca515ea3cc7ddfa11f03938477047f7240f081/src/main/java/org/gleisbelegung/sts/XmlKeys.java) schon mal angefangen. Ich vertröste dich auf nachfolgenden Diff.

manuel3108 replied:
Alles klar. Soll ich den kram dann schonmal mergen, oder einfach offenlassen?

Greonyral replied:
Kannst du offen lassen. Der nächste Diff ist dann merge in diesen Branch. Dann sind die Änderungen in diesen Branch drin, du musst die aber nicht mehr anschauen und die Änderungen hier laufen nicht "halb fertig" in master rein

Greonyral commented on src/main/java/org/gleisbelegung/sts/ScheduleFlagsInterface.java:
```diff
@@ -0,0 +1,24 @@
+public interface ScheduleFlagsInterface {
+
+    @Nullable
+    TrainInterface getE();
```
Analog zu https://github.com/manuel3108/Gleisbelegung/pull/1#discussion_r302196037.
Kommt noch.

manuel3108 commented on src/main/java/org/gleisbelegung/sts/TrainDetails.java:
```diff
@@ -0,0 +1,184 @@
+    public String toString() {
+        return this.name + " " + this.source + " -> " + this.target + " "
+                + this.delay + " "
+                + (!this.visible ? " X" : (this.atPlatform ? " -" : ""));
```
`(this.visible ? (this.atPlatform ? " -" : "") : " X");`
should make the code easier to read.

I.e. this will output `IC 442 C-City -> A-Stadt 0 X`
What should the X say? That the train is not visible?

Greonyral replied:
Yes - It's a copy-n-pasted method from my code. I want to remove the encapsulated conditionals ... on ToDo for one of next change sets.

manuel3108 replied:
Do you want to address this within this PR or later?

manuel3108 commented:
Please also have a look at the result of Codacy.
I already ignored some issues, but those which are open right now should be fixed in my opinion

Greonyral commented:
The ScheduleFlags class have been broken into smaller ones. Testing with real StS shows that `XmlInputHandlerThread.handleSchedule(XML)` causes the connection to break. Removing  `stsSocket.registerEvent(eventType, t);` keeps the connection intact => I will take care of that.

manuel3108 commented:
I think ignoring the test folder is a good idea.
Unfortunately its not possible to get warnings but ignore those in the overall score. The tests are kinda checked (at least if they compile and if they run correctly) by circle ci, so for now we will completly ignore warnings in the tests folder.
This PR just went down from 48 to 33 issues, which is nice

Greonyral commented:
> Testing with real StS shows that XmlInputHandlerThread.handleSchedule(XML) causes the connection to break. Removing stsSocket.registerEvent(eventType, t); keeps the connection intact => I will take care of that.

Fixed by https://github.com/manuel3108/Gleisbelegung/pull/1/commits/fa99a2145ace83f573d9c60b64f41564a4bd4030

Greonyral commented:
I finally found the reason (hopefully) messing up the XML see 
686fd5a 

New Bug found:
```
java.util.NoSuchElementException
	at java.util.ArrayList$ListItr.previous(ArrayList.java:939)
	at org.gleisbelegung.sts.Schedule.updateByXml(Schedule.java:212)
	at org.gleisbelegung.XmlInputHandlerThread.handleSchedule(XmlInputHandlerThread.java:161)
	at org.gleisbelegung.XmlInputHandlerThread.handleInput(XmlInputHandlerThread.java:229)
	at org.gleisbelegung.XmlInputHandlerThread.run(XmlInputHandlerThread.java:258)
```

manuel3108 commented:
Will this possibly also address the problem described in #2?

Greonyral commented:
> 
> 
> Will this possibly also address the problem described in #2?

The error looks identically so possibly yes, and it explain why it was so hard to reproduce. It wasn't a bug it was a race-condition of concurrent sending of XML.

manuel3108 commented:
I tested it, and it seems to be better. But now sometimes (1 out of 10) I get the error message `Gegenstelle antwortet nicht mehr` fired in https://github.com/manuel3108/Gleisbelegung/blob/686fd5aae2d9bd2e38ac26aa0301c3edbabb0f6f/src/main/java/org/gleisbelegung/XmlInputHandlerThread.java#L207

manuel3108 commented:
Ops, forgot to pull. Works great now :D

Greonyral commented:
The code-style issues within the `sts` package should fixed now. There are issues of `XmlStringBuilder` (won't be fixed at #1) and the `plugin-tester` (might be fixed in the mentioned change, to work with recorded real StS games) left

Greonyral commented:
@manuel3108 the CI test failed again, but locally the tests do not fail.
```
org.gleisbelegung.sts.PlatformFactoryTest > testIncomplete FAILED
    java.lang.AssertionError
```

I noticed that the test had some dependency on the other test which leads sometimes to having the Platform 5 defined, where the test expected it's not => fixed in c30ff08

Greonyral commented:
The bugfix for the bug I found mentioned at https://github.com/manuel3108/Gleisbelegung/pull/1#issuecomment-515782559 has been luckily recorded. I added a test with exact that broken update  9db3b12 and added some lines to prevent the bug in the future 58e798f, see #20

manuel3108 commented:
What about the issues caused by codacy? Do you want to solve them, or do you think it is better if I remove on of these rules?
Could you please also merge the master into this branch? Then I will be happy to merge.
